ESSCertIDv2 Update for RFC 3161 (ts module)

kleinmrk opened the pull request:
This patch introduces new .conf variables (ess_cert_id_v2 and ess_cert_id_v2_alg) and adds support for ESSCertIDv2 to ts module as defined in [RFC5816](https://www.rfc-editor.org/rfc/rfc5816.txt), thus it removes another hardcoded SHA-1 usage from ts module.

Original behavior (ESSCertID) is preserved.

kleinmrk commented:
Yes, we do need both since it is written in [RFC5816](https://www.rfc-editor.org/rfc/rfc5816.txt) that the certificate identifier is either ESSCertID or ESSCertIDv2.

richsalz commented on crypto/ts/ts_conf.c:
```diff
@@ -88,6 +88,8 @@
 #define ENV_CLOCK_PRECISION_DIGITS      "clock_precision_digits"
 #define ENV_VALUE_YES                   "yes"
 #define ENV_VALUE_NO                    "no"
+#define ENV_ESS_CERT_ID_V2              "ess_cert_id_v2"
```
do we really need these define's?

kleinmrk replied:
I have simplified the logic and now only one .conf variable is needed.

richsalz commented on crypto/ts/ts_conf.c:
```diff
@@ -517,3 +519,34 @@ int TS_CONF_set_ess_cert_id_chain(CONF *conf, const char *section,
+    const EVP_MD *cert_md = NULL;
+    const char *md = NULL;
+
+    md = NCONF_get_string(conf, section, ENV_ESS_CERT_ID_V2_ALG);
```
merge this up onto line 535

richsalz commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -719,12 +729,27 @@ static int ts_RESP_sign(TS_RESP_CTX *ctx)
-        goto err;
+    if (ctx->flags & TS_ESS_CERT_ID_V2) {
+        certs = ctx->flags & TS_ESS_CERT_ID_CHAIN ? ctx->certs : NULL;
+        if (!
```
break this up into two lines.  sc2 = ... and then if (!sc1 ==  NULL)
dont use ! for pointer tests any more.

richsalz commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -884,6 +910,129 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
+        goto err;
+    for (i = 0; i < sk_X509_num(certs); ++i) {
+        X509 *cert = sk_X509_value(certs, i);
+        if ((cid = ESS_CERT_ID_V2_new_init(hash_alg, cert, 1)) == NULL
```
blank line after var decls.

richsalz commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -884,6 +910,129 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
+    unsigned int hash_len = EVP_MAX_MD_SIZE;
+    X509_ALGOR *alg = NULL;
+
+    memset(hash, 0, EVP_MAX_MD_SIZE);
```
make the last arg be "sizeof(hash)"

richsalz commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -884,6 +910,129 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
+    unsigned char *p, *pp = NULL;
+    int len;
+
+    len = i2d_ESS_SIGNING_CERT_V2(sc, NULL);
```
merge this up to the "int len" line above.  keep the blank line :)

richsalz commented on crypto/ts/ts_rsp_verify.c:
```diff
@@ -282,6 +306,17 @@ static ESS_SIGNING_CERT *ess_get_signing_cert(PKCS7_SIGNER_INFO *si)
+    ASN1_TYPE *attr;
+    const unsigned char *p;
+    attr = PKCS7_get_signed_attribute(si, NID_id_smime_aa_signingCertificateV2);
+    if (!attr)
```
use == NULL as the test.  and put a blank line before.

richsalz commented on crypto/ts/ts_rsp_verify.c:
```diff
@@ -311,6 +346,40 @@ static int ts_find_cert(STACK_OF(ESS_CERT_ID) *cert_ids, X509 *cert)
+    for (i = 0; i < sk_ESS_CERT_ID_V2_num(cert_ids); ++i) {
+        ESS_CERT_ID_V2 *cid = sk_ESS_CERT_ID_V2_value(cert_ids, i);
+
+        const EVP_MD *md;
```
swap this and the blank line above :)

richsalz commented:
a few minor style issues.  I didn't always comment on all of them, so please make sure to check that there's a blank line after variable declarations and don't use ! for pointer checking.
rebase, squash, repost and then I can plus-one this.

kleinmrk commented:
@richsalz thank you for the review.
What do you mean by repost? Should I open a new pull request?

richsalz commented:
No, this MR is fine.  Just go to the branch, rebase on mater and push it to github again (-f flag)

richsalz commented:
+1  will merge after anothe rreview.  thanks.

richsalz commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -884,6 +908,130 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
+    return 0;
+}
+
 static ASN1_GENERALIZEDTIME
```
merge 1035/1306, put the linebreak *after* the paren.

richsalz commented:
any chance of tests?

kleinmrk commented:
I've added ess_cert_id_alg variable into existing tsa tests. I believe this is enough.

richsalz commented:
ok thanks

jariq commented:
It's very good to see TS module is getting some attention - I've been waiting five years for inclusion of [my last patch](https://rt.openssl.org/Ticket/Display.html?id=2145&user=guest&pass=guest) :smiley:. Any chance of getting this merged to 1.1?

richsalz commented:
Yes, the OpenSSL has not care very much about TS.  Sorry about that.
I hope this will make it; it needs another team member to review.

smuehlst commented on include/openssl/ts.h:
```diff
@@ -635,6 +662,7 @@ void ERR_load_TS_strings(void);
 # define TS_F_TS_RESP_CTX_SET_ACCURACY                    128
 # define TS_F_TS_RESP_CTX_SET_CERTS                       129
 # define TS_F_TS_RESP_CTX_SET_DEF_POLICY                  130
+# define TS_F_TS_RESP_CTX_SET_ESS_CERT_ID_V2_DIGEST       159
```
Define TS_F_TS_RESP_CTX_SET_ESS_CERT_ID_V2_DIGEST is not used anywhere in the code.

smuehlst commented on include/openssl/ts.h:
```diff
@@ -660,12 +688,14 @@ void ERR_load_TS_strings(void);
 # define TS_R_BAD_TYPE                                    133
 # define TS_R_CANNOT_LOAD_CERT                            137
 # define TS_R_CANNOT_LOAD_KEY                             138
+# define TS_R_CANNOT_USE_SHA1_WITH_ESSCERTIDV2            140
```
Define TS_R_CANNOT_USE_SHA1_WITH_ESSCERTIDV2 is not used anywhere in the code.

smuehlst commented:
The following two error message defines were added, but are not used anywhere in the code:

- TS_F_TS_RESP_CTX_SET_ESS_CERT_ID_V2_DIGEST
- TS_R_CANNOT_USE_SHA1_WITH_ESSCERTIDV2

mattcaswell commented:
As a new feature it looks like this has missed 1.1.0

richsalz commented:
See also #206

levitte commented:
Time to look at this?

snhenson commented:
TBH It looks as though the ESSCertIDv2 functionality belongs more in CMS than TS. Both CMS and TS could then make use of it.

jariq commented:
> TBH It looks as though the ESSCertIDv2 functionality belongs more in CMS than TS.

I don't say I don't agree but ESSCertID(v1) is currently defined in TS too. Just saying.

kleinmrk commented:
Rebased + removed unused defines pointed out by @smuehlst.
Please let me know if I can do something more to get this merged.

richsalz commented:
and also, consider splitting the ts.pod section on configuration file into man5/tsconfig and updating all the links.  but that's a separate issue., don't do it here.

kleinmrk commented:
@richsalz 
> looks okay to me, but needs an entry for ess_ert_id_alg added to man1/ts.pod

It is already there, have a look at man1/ts.pod line 506. Or do you mean something else?

richsalz commented:
You're right, i missed it.  Thanks.  But do please also remove the "contact Zoltan" section from that page.

kleinmrk commented:
Contact part removed

FdaSilvaYY commented:
FYI, There a bunch of style issue.
Have you use [openssl-format-source](https://github.com/openssl/openssl/blob/master/util/openssl-format-source) ?
If yes, you also need to declare the new structure types in indent.pro ( aside in the same folder).
It will improve the results.

FdaSilvaYY commented on include/openssl/ts.h:
```diff
@@ -156,6 +161,22 @@ ESS_SIGNING_CERT *d2i_ESS_SIGNING_CERT(ESS_SIGNING_CERT **a,
+void ESS_SIGNING_CERT_V2_free(ESS_SIGNING_CERT_V2 * a);
+int i2d_ESS_SIGNING_CERT_V2(const ESS_SIGNING_CERT_V2 * a,
+                            unsigned char **pp);
+ESS_SIGNING_CERT_V2 *d2i_ESS_SIGNING_CERT_V2(ESS_SIGNING_CERT_V2 ** a,
```
Style: ESS_CERT_ID_V2 **a,

FdaSilvaYY commented on include/openssl/ts.h:
```diff
@@ -156,6 +161,22 @@ ESS_SIGNING_CERT *d2i_ESS_SIGNING_CERT(ESS_SIGNING_CERT **a,
 
+ESS_CERT_ID_V2 *ESS_CERT_ID_V2_new(void);
+void ESS_CERT_ID_V2_free(ESS_CERT_ID_V2 * a);
+int i2d_ESS_CERT_ID_V2(const ESS_CERT_ID_V2 * a, unsigned char **pp);
```
Style: TYPE *ptr

FdaSilvaYY commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -835,9 +859,135 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
+static ESS_SIGNING_CERT_V2 *ESS_SIGNING_CERT_V2_new_init(const EVP_MD *hash_alg,
+                                                         X509 *signcert,
+                                                         STACK_OF(X509)
+                                                         *certs)
```
Style: merge this line with previous, it will not exceed 80 cols.

FdaSilvaYY commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -835,9 +859,135 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
+    unsigned char *p, *pp = NULL;
+    int len = i2d_ESS_SIGNING_CERT_V2(sc, NULL);
+
+    if ((pp = (unsigned char *)OPENSSL_malloc(len)) == NULL) {
```
Useless cast, IMHO

kleinmrk commented:
@FdaSilvaYY, I tried to run `./util/openssl-format-source crypto/ts/ts_rsp_sign.c` and also on other files I've edited but that introduced a lot of changes also in parts of code that I did not intend to change. I'm afraid it wouldn't be clear what is the real change and what is refactor. Furthermore, the result was not as I expected even though I have added `-T ESS_CERT_ID_V2`, `-T ESS_SIGNING_CERT_V2` and `-T STACK_OF_ESS_CERT_ID_V2_` into indent.pro. I ended up with many entries like this one:
`STACK_OF (ESS_CERT_ID_V2) * cert_ids;`
I expected something like `STACK_OF(ESS_CERT_ID_V2) *cert_ids;` so I am not sure whether this is okay or not.

I have addressed your comments manually and updated indent.pro for some future refactoring.

If you still think that I should run openssl-format-source just let me know.

richsalz commented:
addressing the comments and putting the types in indent.pro is fine for this PR.

levitte commented:
That failing builds indicate a memory leak.  The build that enables asan might give the most accurate information.

Also, needs another rebase, but that should be a minor job.

kleinmrk commented:
I 've fixed the memory leak.

FdaSilvaYY commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -835,9 +859,132 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
+        goto err;
+
+    if ((cid = ESS_CERT_ID_V2_new_init(hash_alg, signcert, 0)) == NULL
+        || !sk_ESS_CERT_ID_V2_push(sc->cert_ids, cid))
```
_Small remark_: when the `sk_ESS_CERT_ID_V2_push` fails, you will have to delete yourself the unpushed item `cid`.
it is done below with this comment `/* Ownership is lost. */`

FdaSilvaYY commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -835,9 +859,132 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
+        X509 *cert = sk_X509_value(certs, i);
+
+        if ((cid = ESS_CERT_ID_V2_new_init(hash_alg, cert, 1)) == NULL
+            || !sk_ESS_CERT_ID_V2_push(sc->cert_ids, cid))
```
same as above

richsalz commented:
CLA now on file.

FdaSilvaYY commented:
Ping ! Need a quick rebase because of `libcrypto.num` .

kleinmrk commented:
@FdaSilvaYY, rebased.

richsalz commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -835,9 +859,138 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
-static ASN1_GENERALIZEDTIME
-*TS_RESP_set_genTime_with_precision(ASN1_GENERALIZEDTIME *asn1_time,
-                                    long sec, long usec, unsigned precision)
+static ESS_SIGNING_CERT_V2 *ESS_SIGNING_CERT_V2_new_init(const EVP_MD *hash_alg,
```
please use lowercase ESS for static functions.

richsalz commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -835,9 +859,138 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
+
+    if ((sc = ESS_SIGNING_CERT_V2_new()) == NULL)
+        goto err;
+    if (sc->cert_ids == NULL &&
```
sc was just created and either cert_ids is guaranteed initialized, or not.  some part of this if statement isn't needed :)

richsalz commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -835,9 +859,138 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
+    return NULL;
+}
+
+static ESS_CERT_ID_V2 *ESS_CERT_ID_V2_new_init(const EVP_MD *hash_alg,
```
lowercase ESS for static function.

richsalz commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -835,9 +859,138 @@ static int ESS_add_signing_cert(PKCS7_SIGNER_INFO *si, ESS_SIGNING_CERT *sc)
+        goto err;
+
+    if (issuer_needed) {
+        if (cid->issuer_serial == NULL
```
same comment about the init and if statement as above.

richsalz commented on crypto/ts/ts_rsp_verify.c:
```diff
@@ -243,6 +268,18 @@ static ESS_SIGNING_CERT *ess_get_signing_cert(PKCS7_SIGNER_INFO *si)
     return d2i_ESS_SIGNING_CERT(NULL, &p, attr->value.sequence->length);
 }
 
+static ESS_SIGNING_CERT_V2 *ess_get_signing_cert_v2(PKCS7_SIGNER_INFO *si)
```
lowercase ESS.  will stop pointing it out now :)

richsalz commented on crypto/ts/ts_rsp_verify.c:
```diff
@@ -272,6 +309,41 @@ static int ts_find_cert(STACK_OF(ESS_CERT_ID) *cert_ids, X509 *cert)
+    unsigned char cert_digest[EVP_MAX_MD_SIZE];
+    unsigned int len;
+
+    if (cert_ids == NULL || cert == NULL)
```
are NULL valid values, or do they mean programming error?  if programming error, then remove this test.

kleinmrk commented:
@richsalz, I incorporated requested changes.
I have one more question. How should I name internal structures? The [ts_lcl.h](https://github.com/openssl/openssl/blob/master/crypto/ts/ts_lcl.h) file seems to be inconsistent some structures do have the _st suffix and some don't. Should I add the _st suffix to the new structures?

richsalz commented:
Thanks for asking.  Do whatever most do.  We are not consistent with this. :(

levitte commented:
A `_st` suffix is prefered.

kleinmrk commented:
I've added _st suffixes

richsalz commented:
can someone else take a look at this?  It fixes #2119

levitte commented:
For starters, it needs a rebase on top of a fresh master...

levitte commented on crypto/ts/ts_rsp_sign.c:
```diff
@@ -35,7 +35,16 @@ static ESS_SIGNING_CERT *ess_SIGNING_CERT_new_init(X509 *signcert,
+static ESS_CERT_ID_V2 *ess_CERT_ID_V2_new_init(const EVP_MD *hash_alg,
+                                               X509 *cert, int issuer_needed);
+static int ess_add_signing_cert_v2(PKCS7_SIGNER_INFO *si,
+                                   ESS_SIGNING_CERT_V2 *sc);
```
Inconsistent naming...  why capitalize some function names and not others?

levitte commented on doc/man1/ts.pod:
```diff
@@ -605,10 +610,7 @@ You could also look at the 'test' directory for more examples.
-Zoltan Glozik <zglozik@opentsa.org>. Known issues:
-
-=over 2
+=over 4
```
I think we concluded elsewhere that pointed lists should be `=over 2`, so please revert the change on this line.

richsalz commented:
Sigh.  I feel kinda bad for the submittor.  It's like we keep adding things for him to do.

jariq commented:
He'll be rewarded with eternal glory.

kleinmrk commented:
@levitte static functions are now consistently in lowercase.

levitte commented:
Cool!  Going in...

levitte commented:
Merged.  I smashed it all together into one commit, f0ef20bf386b5c37ba5a4ce5c1de9a819bbeffb2
